Feature/issue 864 matern52 prim

drezap opened the pull request:
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Add matern 5/2 kernel to prim. This is consistent with GPML, but includes a marginal standard deviation. This also had ARD support. 

I opened a new PR to include the proper issue, but also includes updated function names, as well as addresses issues brought up in another matern pull request (more efficient calculation of ARD and updating to `check_finite_postive`)

#### How to Verify:

Testing:
All combinations of double, vectors and row vectors, and error handling.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Copyright 2018 Andre Zapico

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

drezap commented:
can some one take a look at this? I'm hitting the exponential hard tomorrow. I tried to include the comments from that one here

syclik commented:
@drezap: the copyright holder is required. Please indicate who owns the copyright. (typically your employer or yourself.)

syclik commented:
Thanks. It doesn't show, so I think whatever you're typing may be commenting things out. 

> On May 23, 2018, at 11:23 AM, drezap <notifications@github.com> wrote:
> 
> When I write copyright 2018 my name, the last part in brackets doesn’t show up when i publish the comment, so if you see copyright 2018 I wrote my name. I can fix when I’m at a computer if i forgot altogether sorry!
> 
> > On May 23, 2018, at 6:02 PM, Daniel Lee <notifications@github.com> wrote:
> > 
> > @drezap: the copyright holder is required. Please indicate who owns the copyright. (typically your employer or yourself.)
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+  return cov;
+}
+
+/** Returns a Matern 5/2 covariance matrix with one input vector
```
Same doc comments as above.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+ *  \frac{\sqrt{5}d(x, x')}{l})exp(-\frac{5 d(x, x')^2}{3l^2})
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the euclidean distance.
```
Typo: capitalize "Euclidean"

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern52_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const std::vector<T_l> &length_scale) {
+  using stan::math::sqrt;
```
remove the `using stan::math::*` statements.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      temp = 0;
```
Shift the declaration of `temp` to here.

syclik replied:
We should be able to move the computation of `temp` to outside these for loops and reuse it.

drezap replied:
can't do this, needs to reset for each element in the matrix, otherwise it would be a cumulative ARD sum of all elements in the matrix, which is invalid... needs to be for each element in the covariance matrix. everything I'm thinking amounts to more steps than this..

syclik replied:
I don't follow. It looks like this term is used `n * (n - 1) / 2` times (pseudocode): `temp = sum(1 / length_scale[1 : l_size]^2)`

I would think that computation could go outside the loop. Did I miss something?

drezap replied:
You are correct

drezap replied:
you're correct on the implementation/optimization, but I'm try to convince myself that this is according to as people define ARD: "seperate length scale for each argument" and in the matlab/GPML document below, as they are index by `m` as well (or k in my implementation)

drezap commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+/** Returns a Matern 5/2 covariance matrix with one input vector
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{5}d(x, x')}{l})exp(-\frac{5 d(x, x')^2}{3l^2})
```
totally butchered these, although I implemented correctly. See updates on next commit

drezap commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,273 @@
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_sq
+                  * (1.0 + root_5_inv_l * sqrt(squared_distance(x[i], x[j]))
```
cool  I'm updating and using the following rule: sacrifice a bit of memory instead of calling a function more than once

drezap commented:
I see what the issue is with ARD priors. Here, we only implemented the covariance functions for 1D. So the length of the ‘length_scale’ should be 1... i don’t think it make sense to just sum over some length scale. So I need to open up an issue and extend these PRs to be multidimensional, (i.e take an Eigen::Matrix input). I will open an issue, and I need to fix the ARD that already went into ‘cov_exp_quad’. I’ll run it by someone here to make sure it’s accurate, but I think I’m correct

drezap commented:
what I said was incorrect, they do take multivariate input, but the summation is over the inner vectors, so I may have to do some meta programming to do this properly don't think I can use the `squared_distance`, but i'll think about it tomorrow when it's not so late.

drezap commented:
how do I reset this travis build? I've reset the jenkins. looks like this isn't "not passing" because of stuff I've implemented...

drezap commented:
closing until ready

drezap commented:
this matches output of other implementations of this kernel...

drezap commented:
@syclik

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type of elements contained in vector x
+ * @tparam type of element of sigma, marginal standard deviation
```
This comment is wrong. This should have been caught by doxygen, but `type` should be `T_s`.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5_inv_l = pow(5.0, 0.5) / length_scale;
```
use `sqrt(5)` instead of `pow(5.0, 0.5)`.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+  T_s sigma_sq = square(sigma);
+  T_l root_5_inv_l = pow(5.0, 0.5) / length_scale;
+  T_l inv_l_sq_5_3 = 5.0 / (3.0 * square(length_scale));
+  T_l neg_root_5_inv_l = -1.0 * pow(5.0, 0.5) / length_scale;
```
use `-sqrt(5)` for clarity.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+  T_l root_5_inv_l = pow(5.0, 0.5) / length_scale;
+  T_l inv_l_sq_5_3 = 5.0 / (3.0 * square(length_scale));
+  T_l neg_root_5_inv_l = -1.0 * pow(5.0, 0.5) / length_scale;
+  typename return_type<T_x>::type sq_distance;
```
I think there's a more appropriate template meta program to use here. I think `scalar_type<T_x>::type` is the right one.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+  for (size_t i = 0; i < x_size; ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      sq_distance = squared_distance(x[i], x[j]);
```
Can you move the declaration of `sq_distance` and `distance` here? There's no reason for it to sit outside the loop.

drezap replied:
I'm sorry - do you mean, move the `typename return_type<T_x>::type sq_distance;` inside the loop? is it more expensive to overwrite memory or does it cost more to re-declare something?

syclik replied:
Yes. C++ compilers are very good at this sort of optimization. It might be better because it’s able to drop the variable out of scope.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type of elements contained in vector x
+ * @tparam type of element of sigma, marginal standard deviation
```
replace `type` with `T_s`

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = pow(5.0, 0.5);
```
lots of the same comments from above apply here

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+  T_l root_5 = pow(5.0, 0.5);
+  T_l five_thirds = 5.0 / 3.0;
+  T_l neg_root_5 = -1.0 * pow(5.0, 0.5);
+  typename return_type<T_x>::type sq_distance;
```
As a general rule, move declarations of variables close to their use.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+    }
+  }
+
+  for (size_t i = 0; i < x_size; ++i) {
```
Since this loop is close to the one above, is there any reason this function isn't set up to construct the `x_new` and then call the other function? (It might need a little more work than that, but I think we can remove a lot of duplication.)

drezap replied:
@syclik good catch, yeah this is bad practice. I'm working on a functino in this PR at least in prim that allows to divide by the length scale, and it will be used in all the "special case of matern kernels"

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+ * @tparam T_l type of elements in vector of length scale
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
```
Please document that the length scales have to match the size of dimensions.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,323 @@
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
```
Please update when it throws.

drezap commented:
@syclik can you check this one out? ignore the others, as I'll carry on all comments from here to those ones.

The additions include: addressing your comments, and a function to do the division by length scale instead of the copy paste nonsense

syclik commented on stan/math/prim/mat/fun/divide_columns.hpp:
```diff
@@ -0,0 +1,71 @@
+ * @tparam R   number of rows in the submatrix of x
+ * @tparam C   number of columns in the submatrix of x
+ *
+ * @param x std::vector of elements representing a matrix
```
Please change this comment. The elements are all matrices, not that the collection of the vector represents a single matrix in some way.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern52_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
```
This error message contradicts the doc: should it be "standard deviation"?

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+    for (size_t j = i + 1; j < x_size; ++j) {
+      typename return_type<T_x>::type sq_distance
+          = squared_distance(x[i], x[j]);
+      typename return_type<T_x>::type root_sq_distance = sqrt(sq_distance);
```
why not call this variable `distance`?

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = sqrt(5.0);
+  T_l five_thirds = 5.0 / 3.0;
```
this is overpromoted. Just make this a `double`.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = sqrt(5.0);
```
this is overpromoted.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = sqrt(5.0);
+  T_l five_thirds = 5.0 / 3.0;
+  T_l neg_root_5 = -sqrt(5.0);
```
same here. Over promotion.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+    for (size_t j = 0; j < x2_size; ++j) {
+      typename return_type<T_x1, T_x2>::type sq_distance
+          = squared_distance(x1[i], x2[j]);
+      typename return_type<T_x1, T_x2>::type root_sq_distance
```
same here: why not call it `distance`?

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+  for (size_t n = 0; n < x2_size; ++n)
+    check_not_nan("gp_matern52_cov", "x1", x2[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
```
I think this is standard deviation. Please verify and make everything consistent.

drezap replied:
I'm just using magnitude for these from now on, as this is consistent with GP literature

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = sqrt(5.0);
```
Over promotion. Please just use `double`.

syclik commented on stan/math/prim/mat/fun/gp_matern52_cov.hpp:
```diff
@@ -0,0 +1,302 @@
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = sqrt(5.0);
+  T_l five_thirds = 5.0 / 3.0;
+  T_l neg_root_5 = -sqrt(5.0);
```
Just use `-root_5` here or just replace the use with `-root_5`.

syclik commented on test/unit/math/prim/mat/fun/gp_matern52_cov_test.cpp:
```diff
@@ -0,0 +1,1398 @@
+  msg2 = pull_msg(x, sigma_bad, l);
+  msg3 = pull_msg(x, sigma_bad, l_bad);
+  EXPECT_TRUE(std::string::npos != msg1.find(" length scale")) << msg1;
+  EXPECT_TRUE(std::string::npos != msg2.find(" marginal variance")) << msg2;
```
These might need to change.

syclik commented on test/unit/math/prim/mat/fun/gp_matern52_cov_test.cpp:
```diff
@@ -0,0 +1,1398 @@
+               std::invalid_argument);
+}
+
+TEST(MathPrimMat, blahblah) {
```
`blahblah`? I don't think this is meant to be here based on the content of the test. Please remove if not useful.

drezap commented:
I've updated this... @syclik regarding this: https://github.com/stan-dev/math/pull/945...
I get from the autodiff paper for the `rev` version of this stuff you're using PImpl/reading a reference for Eigen::Matrix, do you want me to carry this over to prim from now on? I was just following the the initial cov_exp_quad implementation, should I always just reference when modifying Eigen::Matrices?

drezap commented:
@syclik can you have a look please? Some of the others are waiting for this to go through.. thanks!!

syclik commented:
Can you remind me after we have fixes for the makefiles? That has priority
for the moment.

On Thu, Jul 26, 2018 at 3:10 AM Andre Zapico <notifications@github.com>
wrote:

> @syclik <https://github.com/syclik> can you have a look please? Some of
> the others are waiting for this to go through.. thanks!!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/867#issuecomment-407999369>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_F4sWkLhd7N0PvE935NhfkH9Ijkd_ks5uKWtCgaJpZM4UG0FE>
> .
>

syclik commented:
I'm not sure what you mean by this:
> I get from the autodiff paper for the rev version of this stuff you're using PImpl/reading a reference for Eigen::Matrix, do you want me to carry this over to prim from now on? I was just following the the initial cov_exp_quad implementation, should I always just reference when modifying Eigen::Matrices?
> 
> i.e. b->coefReff(i, j)

What's the alternative? Just using `b->(i,j)`?

drezap commented:
@syclik could you pass this off to @rok-cesnovar or something? I've already made the required changes, and with this PR I could put some of the others I have open out  (I have a function in this PR that I'll then include in those ones). It doesn't touch the autodiff stack, so I think this is something someone with less autodiff experience can handle, and it would take some weight off of your shoulders.
